### PR TITLE
[Feature](Nereids) Support join hint.

### PR DESCRIPTION
**Pull request by wangshuo128**
# Proposed changes

Support join hint for nereids planner.  Hints for `broadcast` and `shuffle` are supported by this PR.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 1]
```diff
@@ -851,10 +855,23 @@ private LogicalPlan withJoinRelations(LogicalPlan input, RelationContext ctx) {
...
 
+            JoinHint joinHint = Optional.ofNullable(join.joinHint()).map(hintCtx -> {
+                String hint = typedVisit(join.joinHint());
+                if (JoinHint.JoinHintType.SHUFFLE.toString().equalsIgnoreCase(hint)) {
+                    return JoinHint.SHUFFLE_RIGHT;
+                } else if (JoinHint.JoinHintType.BROADCAST.toString().equalsIgnoreCase(hint)) {
+                    return JoinHint.BROADCAST_RIGHT;
+                } else {
+                    throw new ParseException("Invalid join hint: " + hint, hintCtx);
+                }
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:866]: we can add them into a function like `public static JoinHint of(String name)`

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.4 seconds
 load time: 441 seconds
 storage size: 17148662222 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221115113631_clickbench_pr_46389.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java [hunk 2]
```diff
@@ -90,6 +91,11 @@ public Plan joinToMultiJoin(Plan plan) {
             return plan;
         }
 
+        // Can't reorder the join inputs if this join has hints.
+        if (plan instanceof LogicalJoin && ((LogicalJoin) plan).hasJoinHint()) {
+            return plan;
+        }
+
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:98]: could we still eliminate all cross join if we add this?

File: fe/fe-core/src/test/java/org/apache/doris/nereids/util/ExceptionChecker.java [hunk 3]
```diff
@@ -37,6 +37,7 @@ public ExceptionChecker assertMessageEquals(String message) {
     }
 
     public ExceptionChecker assertMessageContains(String context) {
+        System.out.println(exception.getMessage());
         Assertions.assertTrue(exception.getMessage().contains(context));
```
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/util/ExceptionChecker.java:41]: Assertions.assertTrue(exception.getMessage().contains(context), exception.getMessage());

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 [hunk 4]
```diff
@@ -122,7 +122,13 @@ relation
     ;
 
 joinRelation
-    : (joinType) JOIN right=relationPrimary joinCriteria?
+     : (joinType) JOIN joinHint? right=relationPrimary joinCriteria?
+     ;
```
Comment (morrySnow) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:126]:
```suggestion
    : (joinType) JOIN joinHint? right=relationPrimary joinCriteria?
    ;
```

**Review by morrySnow — APPROVED** [review]

**Pull request merged**